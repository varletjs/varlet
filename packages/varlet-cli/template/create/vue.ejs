<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
