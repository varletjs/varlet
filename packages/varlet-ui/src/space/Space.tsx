import { defineComponent, VNodeChild, Fragment, VNode, Comment } from 'vue'
import { internalSizeValidator, props, type SpaceSize } from './props'
import { isArray } from '@varlet/shared'
import { call, createNamespace } from '../utils/components'
import { toSizeUnit } from '../utils/elements'
import { computeMargin } from './margin'
import '../styles/common.less'
import './space.less'

const { n, classes } = createNamespace('space')

export default defineComponent({
  name: 'VarSpace',
  props,
  setup(props, { slots }) {
    const getSize = (size: SpaceSize, isInternalSize: boolean): string[] => {
      return isInternalSize
        ? [`var(--space-size-${size}-y)`, `var(--space-size-${size}-x)`]
        : isArray(size)
        ? (size.map(toSizeUnit) as string[])
        : ([toSizeUnit(size), toSizeUnit(size)] as string[])
    }

    const padStartFlex = (style: string | undefined) => {
      return style === 'start' || style === 'end' ? `flex-${style}` : style
    }

    return () => {
      const { inline, justify, align, wrap, direction, size } = props
      let children: VNodeChild[] = call(slots.default) ?? []
      const isInternalSize = internalSizeValidator(size)
      const [y, x] = getSize(size, isInternalSize)

      const flatten = (vNodes: any) => {
        const result: VNode[] = []

        vNodes.forEach((vNode: any) => {
          if (vNode.type === Comment) return

          if (vNode.type === Fragment && isArray(vNode.children)) {
            vNode.children.forEach((item: VNode) => {
              result.push(item)
            })
            return
          }

          result.push(vNode)
        })

        return result
      }

      children = flatten(children)

      const lastIndex = children.length - 1
      const spacers = children.map((child, index) => {
        const margin = computeMargin(y, x, {
          direction,
          justify,
          index,
          lastIndex,
        })

        return <div style={{ margin }}>{child}</div>
      })

      return (
        <div
          class={classes(n(), n('$--box'), [inline, n('--inline')])}
          style={{
            flexDirection: direction,
            justifyContent: padStartFlex(justify),
            alignItems: padStartFlex(align),
            flexWrap: wrap ? 'wrap' : 'nowrap',
            margin: direction === 'row' ? `calc(-1 * ${y} / 2) 0` : undefined,
          }}
        >
          {spacers}
        </div>
      )
    }
  },
})
