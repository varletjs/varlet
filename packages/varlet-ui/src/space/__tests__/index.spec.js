import VarSpace from '../Space'
import Space from '..'
import { mount } from '@vue/test-utils'
import { createApp, Fragment, h } from 'vue'

test('test space use', () => {
  const app = createApp({}).use(Space)
  expect(app.component(Space.name)).toBeTruthy()
})

describe('test space component props', () => {
  test('test space align', () => {
    ;['stretch', 'center', 'start', 'end', 'baseline'].forEach((align) => {
      const wrapper = mount(VarSpace, {
        props: { align },
      })

      expect(wrapper.find('.var-space').attributes('style')).toContain(
        'align-items: ' + (align === 'start' || align === 'end' ? `flex-${align}` : align)
      )
      wrapper.unmount()
    })
  })

  test('test space justify', () => {
    ;['start', 'end', 'center', 'space-around', 'space-between'].forEach((justify) => {
      const wrapper = mount(VarSpace, {
        props: { justify },
        slots: {
          default: () => h(Fragment, [h('div', 'child'), h('div', 'child'), h('div', 'child')]),
        },
      })

      expect(wrapper.find('.var-space').attributes('style')).toContain(
        'justify-content: ' + (justify === 'start' || justify === 'end' ? `flex-${justify}` : justify)
      )
      expect(wrapper.html()).toMatchSnapshot()
      wrapper.unmount()
    })
  })

  test('test space size', () => {
    ;[
      { sizeType: 'mini', sizeVal: -2 },
      { sizeType: 'small', sizeVal: -3 },
      { sizeType: 'normal', sizeVal: -4 },
      { sizeType: 'large', sizeVal: -6 },
    ].forEach((size) => {
      const wrapper = mount(VarSpace, {
        props: { size: size.sizeType },
      })

      expect(wrapper.find('.var-space').attributes('style')).toContain(`margin: ${size.sizeVal}px 0px;`)
      wrapper.unmount()
    })
  })

  test('test space wrap', async () => {
    const wrapper = mount(VarSpace, {
      props: {
        wrap: true,
      },
    })

    expect(wrapper.find('.var-space').attributes('style')).toContain('flex-wrap: wrap;')
    await wrapper.setProps({ wrap: false })
    expect(wrapper.find('.var-space').attributes('style')).toContain('flex-wrap: nowrap;')
    wrapper.unmount()
  })

  test('test space direction', async () => {
    const wrapper = mount(VarSpace, {
      props: {
        direction: 'row',
      },
    })

    expect(wrapper.find('.var-space').attributes('style')).toContain('flex-direction: row;')
    await wrapper.setProps({ direction: 'column' })
    expect(wrapper.find('.var-space').attributes('style')).toContain('flex-direction: column;')
    wrapper.unmount()
  })

  test('test space inline', async () => {
    const wrapper = mount(VarSpace, {
      props: {
        inline: true,
      },
    })

    expect(wrapper.find('.var-space--inline').exists()).toBe(true)
    await wrapper.setProps({ inline: false })
    expect(wrapper.find('.var-space--inline').exists()).toBe(false)
    wrapper.unmount()
  })
})

test('test space default slots', () => {
  const wrapper = mount(VarSpace, {
    slots: {
      default: 'This is default slots',
    },
  })

  expect(wrapper.find('.var-space').html()).toContain('This is default slots')
  wrapper.unmount()
})
