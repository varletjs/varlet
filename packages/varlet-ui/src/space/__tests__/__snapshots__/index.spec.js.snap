// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test space component props test space justify 1`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: flex-start; flex-wrap: wrap; margin: -4px 0px;\\">
  <div style=\\"margin: 4px 12px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 12px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 0px;\\">
    <div>child</div>
  </div>
</div>"
`;

exports[`test space component props test space justify 2`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: flex-end; flex-wrap: wrap; margin: -4px 0px;\\">
  <div style=\\"margin: 4px 12px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 12px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 0px;\\">
    <div>child</div>
  </div>
</div>"
`;

exports[`test space component props test space justify 3`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: center; flex-wrap: wrap; margin: -4px 0px;\\">
  <div style=\\"margin: 4px 12px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 12px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 0px;\\">
    <div>child</div>
  </div>
</div>"
`;

exports[`test space component props test space justify 4`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: space-around; flex-wrap: wrap; margin: -4px 0px;\\">
  <div style=\\"margin: 4px 6px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 6px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 6px;\\">
    <div>child</div>
  </div>
</div>"
`;

exports[`test space component props test space justify 5`] = `
"<div class=\\"var-space var--box\\" style=\\"flex-direction: row; justify-content: space-between; flex-wrap: wrap; margin: -4px 0px;\\">
  <div style=\\"margin: 4px 6px 4px 0px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 6px;\\">
    <div>child</div>
  </div>
  <div style=\\"margin: 4px 0px 4px 6px;\\">
    <div>child</div>
  </div>
</div>"
`;
